### PR TITLE
6616245: NullPointerException when using JFileChooser with a custom FileView

### PR DESCRIPTION
**Pull request by TejeshR13**
When a custom `FileView` is used and folder traversal is restricted to a particular directory NPE occurs when user tries to traverse/select other folders except traversable folder. This is caused because when user selects folder other than traversable, the traversal is rejected and hence no file is selected as `currentDirectory` of `JFileChooser`. When user tries to access the restricted folder second time, previous selected file check is failing because of NPE since `getFileChooser().getCurrentDirectory();` is null. To fix the issue, NPE check is added.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-6616245](https://bugs.openjdk.org/browse/JDK-6616245): NullPointerException when using JFileChooser with a custom FileView

### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * @kumarabhi006 (no known github.com user name / role) ⚠️ Review applies to [f5d345e5](https://git.openjdk.org/jdk/pull/10485/files/f5d345e553d1d1acafe1bdc09fde62c85581b3a8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10485/head:pull/10485` \
`$ git checkout pull/10485`

Update a local copy of the PR: \
`$ git checkout pull/10485` \
`$ git pull https://git.openjdk.org/jdk pull/10485/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10485`

View PR using the GUI difftool: \
`$ git pr show -t 10485`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10485.diff">https://git.openjdk.org/jdk/pull/10485.diff</a>

</details>

### CODE REVIEW

**Comment by kumarabhi006** [comment]
The instruction are not quite clear. Can you please modify it as we discussed over call?

**Comment by TejeshR13** [comment]
> The instruction are not quite clear. Can you please modify it as we discussed over call?

Updated.

**Comment by TejeshR13** [comment]
@kumarabhi006 - Reviewer

**Comment by TejeshR13** [comment]
@azuev-java Can you please review this PR.

File: src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java [hunk 1]
```diff
@@ -1266,7 +1266,9 @@ protected DirectoryComboBoxAction() {
         public void actionPerformed(ActionEvent e) {
             directoryComboBox.hidePopup();
             File f = (File)directoryComboBox.getSelectedItem();
-            if (!getFileChooser().getCurrentDirectory().equals(f)) {
+            File curDir = getFileChooser().getCurrentDirectory();
+
+            if (curDir != null && !curDir.equals(f)) {
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: If `f` is always not null, you can you use `!f.equals(curDir)`.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    public CustomFileView(String path) {
+        basePath = path;
+    }
+
+    public Boolean isTraversable(File filePath) {
+        if ((filePath != null) && (filePath.isDirectory())) {
+            return filePath.getAbsolutePath().startsWith(basePath);
+        } else {
+            return false;
+        }
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:114]: I can be simplified to:
```suggestion
        return ((filePath != null) && (filePath.isDirectory())
                && filePath.getAbsolutePath().startsWith(basePath));
```
I'm not insisting though.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                folder again.
+                3. If NPE occurs on second selection then test FAILS, else test
+                is PASS.
+                """;
+        frame = new JFrame("JFileChooser File View NPE test");
+        passFailJFrame = new PassFailJFrame("Test Instructions", INSTRUCTIONS, 5L, 5, 40);
+        jfc = new JFileChooser();
+        String path = "";
+        if (Platform.isWindows()) {
+            path = "C:" + File.separator + "temp";
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:84]: It's rare but possible that Windows is installed not on `C:` drive.

**Review by aivanov-jdk — COMMENTED** [review]
You say that the current directory in `JFileChooser` gets set to `null` because of forbidden navigation to a folder which is not traversable. You avoid the NPE by a null-check.

However, you should prevent the current directory to be set to `null`. If navigation is not allowed, the current directory should remain unchanged _instead of being reset to `null`_.

File: src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java (hunk 1 above)
Comment (TejeshR13) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: > 

Prevention of not setting to null means I have to modify `JFileChooser` class itself, so just handling here. And yeah` f `can't be null, can do that.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 2 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:114]: Can be done, no issues.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 3 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:84]: Since temp will be better option without creating any new directory for the test, I didn't change it (it was reused from JBS report). Any other alternative......?

File: src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java (hunk 1 above)
Comment (aivanov-jdk) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: Why is it bad to modify `JFileChooser` class itself?

I think setting the current directory to `null` while it displays a directory kind of breaks the invariant.

`JFileChooser` navigated to a directory and the current directory is set to it. You tried navigating to another directory, `FileSystem` didn't allow the navigation, so `JFileChooser` continues to display the files in the directory that is showed before. Thus, the current directory hasn't changed. From this point of view, the current directory should remain unchanged.

Comment (TejeshR13) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L608
Here the above logic checks if the current selected directory is traversable and if not it goes up to parent directory till top level and when every check fails then null is set to `currentDirectory`. Should we think of modifying this logic to not to set to null, rather remain in previously valid folder......?

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 3 above)
Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:84]: Why not request the path to the system temporary directory?

Temporary directory on Windows is not located in `C:\temp`. Likely this folder does not exist. You should use `TMP` or `TEMP` environment variable. Alternatively, you can use `createTempFile` from from [`java.io.File`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/File.html#createTempFile(java.lang.String,java.lang.String)) or [`java.nio.Files`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/nio/file/Files.html#createTempFile(java.lang.String,java.lang.String,java.nio.file.attribute.FileAttribute...)) to create a file in the temp directory, its parent is the path to the temp directory.

Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:84]: User's home could be a better option. It's guaranteed to exist, `user.home` system property points to it. And it's a more real-life scenario.

Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:84]: Then I can just make user.home directory as the only traversable.......?

File: src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java (hunk 1 above)
Comment (aivanov-jdk) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: Well, the logic there tries to prevent the current directory from becoming `null`. Yet it traverses up the tree:
https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L603-L607

Probably, if `dir` becomes `null` in the end, it should resort to its default behaviour: `getFileSystemView().getDefaultDirectory()`. Or not change at all?

Then, you can handle the situation here: if `getFileChooser().getCurrentDirectory()` returns `null`, restore the previous current directory.

And what about other L&F? If the same code runs in Windows L&F, Nimbus L&F, any other?

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 3 above)
Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:84]: Makes sense to me.

File: src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java (hunk 1 above)
Comment (TejeshR13) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: In other L&F the same problem exist, but NPE is does not occurs because previous state check is not done, directly the clicked file is been set. 
https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/WindowsFileChooserUI.java#L1328

Comment (aivanov-jdk) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: > Well, the logic there tries to prevent the current directory from becoming `null`. Yet it traverses up the tree:
> 
> https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L603-L607
> 
> Probably, if `dir` becomes `null` in the end, it should resort to its default behaviour: `getFileSystemView().getDefaultDirectory()`. Or not change at all?

This behaviour is specified. https://github.com/openjdk/jdk/blob/ccc1d3169691d066c08e294f5d989b007bfab114/src/java.desktop/share/classes/javax/swing/JFileChooser.java#L576-L580

Yet it doesn't take into account a custom implementation of `FileSystem` which does not allow navigating up. It may be a problem in `FileSystem` implementation.

Comment (TejeshR13) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: In other L&F, null gets assigned to currentDirectory and NPE doesn't occur since that previous directory check is not done.

Comment (aivanov-jdk) [reply src/java.desktop/share/classes/javax/swing/plaf/metal/MetalFileChooserUI.java:1271]: > … NPE doesn't occur since that previous directory check is not done.

Then I suggest going this way and just drop the checks from Metal L&F.

The test should set Metal L&F explicitly. Or better yet, it should be run in all L&Fs to ensure NPE isn't thrown.

<hr>

As for preserving the current value of `currentDirectory` if the passed in folder cannot be navigated to. I think it should be done yet it's a separate issue from this one. Shall I submit a new bug? Does anyone have an opinion on this?

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import javax.swing.SwingUtilities;
+import javax.swing.WindowConstants;
+
+import javax.swing.filechooser.FileView;
+import jdk.test.lib.Platform;
+/*
+ * @test
+ * @bug 6616245
+ * @key headful
+ * @requires (os.family == "windows" | os.family == "linux")
```
[the remark this answers is not in the document]
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:40]: Yeah, was not able to reproduce in macOS.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 3 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 2 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

**Comment by TejeshR13** [comment]
@aivanov-jdk I have updated as per the review comments. Please let me know if its fine. I have put up the mach5 test link in JBS for reference.

**Comment by kumarabhi006** [comment]
@TejeshR13 Verified latest fix on ubuntu and fix is working fine.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        passFailJFrame.awaitAndCheck();
+    }
+
+    static void initialize() throws InterruptedException, InvocationTargetException {
+        JFrame frame;
+        JFileChooser jfc;
+
+        //Initialize the components
+        final String INSTRUCTIONS = """
+                Instructions to Test:
```
Comment (mrserb) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: JFileChooser extends JComponent, probably we can create a frame with embedded JFileChooser and programmatically change the value of that "Look-In ComboBox"-> and catch an exception.

Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: Means in _Test INSTRUCTIONS_ a link to `Look-In` can be attached......?

Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: Or suggesting to make the test automatic......?

Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: @mrserb is suggesting making the test automatic. `JFileChooser` is embedded in a frame in the test. It should be possible… I'm fine with manual test though… for now.

Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: Ok, since the test has to select non traversable folders in both windows and linux I had thought that making it as manual would be better.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                 non-traversable folder from "Look-In" ComboBox by mouse
+                 click/key press. (The folder will not be opened since its
+                 non-traversable). Select the same folder again.
+                3. If NPE occurs on second selection then test FAILS, else test
+                 is PASS.
+                """;
+        frame = new JFrame("JFileChooser File View NPE test");
+        passFailJFrame = new PassFailJFrame("Test Instructions", INSTRUCTIONS, 5L, 13, 40);
+        jfc = new JFileChooser();
+        String path = System.getProperty("user.home");
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:81]: The test does not work for me on Windows. It starts with `user.home` as the current folder as expected; when select `C:\`, it goes to Desktop which is the root of the Shell namespace.

I guess, you have to change it to `System.getProperty("user.home") + File.separator + "Documents"` as you and I discussed. It'll make Desktop unreachable.

On Linux, the test works correctly.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
+
+        frame.add(jfc, BorderLayout.CENTER);
+        frame.pack();
+        frame.setVisible(true);
+    }
+}
+
+class CustomFileView extends FileView {
+    private String basePath;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:97]:
```suggestion
    private final String basePath;
```

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    public CustomFileView(String path) {
+        basePath = path;
+    }
+
+    public Boolean isTraversable(File filePath) {
+        return ((filePath != null) && (filePath.isDirectory())) &&
+                filePath.getAbsolutePath().startsWith(basePath);
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:105]:
```suggestion
        return ((filePath != null) && (filePath.isDirectory()))
                && filePath.getAbsolutePath().startsWith(basePath);
```
I prefer wrapping before the operator, it makes it clear that it's a continuation line. Java Coding Style suggests wrapping this way.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 5 above)
Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: Yet you control which folders are non-traversable. Anyway… A test is better than no test; an automatic test is better than a manual one.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 8 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 7 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+import java.io.File;
+import java.lang.reflect.InvocationTargetException;
+
+import javax.swing.JFrame;
+import javax.swing.JFileChooser;
+import javax.swing.SwingUtilities;
+import javax.swing.WindowConstants;
+
+import javax.swing.filechooser.FileView;
+import jdk.test.lib.Platform;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:35]: It's not used any more, please remove.

Leave one blank line after the last import and the following comment with the jtreg tags.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:40]: You can set Metal L&F explicitly on macOS.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+/*
+ * @test
+ * @bug 6616245
+ * @key headful
+ * @requires (os.family == "windows" | os.family == "linux")
+ * @library /java/awt/regtesthelpers /test/lib
+ * @build PassFailJFrame jdk.test.lib.Platform
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:42]:
```suggestion
 * @library /java/awt/regtesthelpers
 * @build PassFailJFrame
```

`Platform` class from `/test/lib` is not used any more.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            }
+        });
+        passFailJFrame.awaitAndCheck();
+    }
+
+    static void initialize() throws InterruptedException, InvocationTargetException {
+        JFrame frame;
+        JFileChooser jfc;
+
+        //Initialize the components
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:65]: I believe the comment is redundant. Being placed before the instruction text, it's confusing. The method name `initialize` is self-explanatory.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                1. The traversable folder set is to "user/home" Directory,
+                 other than these folder all other
+                 folders are considered as non-traversable.
+                2. When file chooser appears on screen, select any
+                 non-traversable folder from "Look-In" ComboBox by mouse
+                 click/key press. (The folder will not be opened since its
+                 non-traversable). Select the same folder again.
+                3. If NPE occurs on second selection then test FAILS, else test
+                 is PASS.
+                """;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:77]: I propose the following text:

```
                Instructions to Test:
                1. The traversable folder is set to the Documents folder,
                 if it exists, in the user's home folder, otherwise
                 it's the user's home. Other folders are non-traversable.
                2. When the file chooser appears on the screen, select any
                 non-traversable folder from "Look-In" combo box,
                 for example the user's folder or a folder above it.
                 (The folder will not be opened since it's non-traversable).
                3. Select the Documents folder again.
                4. If NullPointerException occurs in the step 3, click Fail,
                 otherwise click Pass.
```
It consistently uses _“folder”_, no mention of “directory”. It explains that the user's home could be the root of the traversable tree. The instructions use imperative to select either _Fail_ or _Pass_.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+        frame = new JFrame("JFileChooser File View NPE test");
+        passFailJFrame = new PassFailJFrame("Test Instructions", INSTRUCTIONS, 5L, 13, 40);
+        jfc = new JFileChooser();
+        String path = System.getProperty("user.home") + File.separator + "Documents";
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:81]:
```suggestion

        String userHome = System.getProperty("user.home");
        String docs = userHome + File.separator + "Documents";
        String path = (new File(docs).exists()) ? docs : userHome;
```
This will make the test more robust, it can continue to run even if `Documents` folder does not exist.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+        jfc.setCurrentDirectory(new File(path));
+        jfc.setFileView(new CustomFileView(path));
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:84]:
```suggestion
        jfc.setFileView(new CustomFileView(path));
        jfc.setControlButtonsAreShown(false);
```
Hide _Open_ and _Cancel_ buttons in the file chooser.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 10 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:42]: In mac the issue was not reproducible.....

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 14 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:87]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 13 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 12 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: 4th point has been modified, since user can only select Pass, otherwise the test itself fails.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 11 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:65]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 9 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

**Review by aivanov-jdk — APPROVED** [review]
Looks good to me now.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 15]
```diff
@@ -65,23 +65,27 @@ static void initialize() throws InterruptedException, InvocationTargetException
...
+                 (The folder will not be opened since it's non-traversable).
+                3. Select the Documents folder again.
+                4. If NullPointerException does not occurs in the step 3,
+                 click Pass, otherwise the Test Fails.
                 """;
         frame = new JFrame("JFileChooser File View NPE test");
         passFailJFrame = new PassFailJFrame("Test Instructions", INSTRUCTIONS, 5L, 13, 40);
         jfc = new JFileChooser();
-        String path = System.getProperty("user.home") + File.separator + "Documents";
+        String userHome = System.getProperty("user.home");
```
[the remark this answers is not in the document]
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+                1. The traversable folder is set to the Documents folder,
+                 if it exists, in the user's home folder, otherwise
+                 it's the user's home. Other folders are non-traversable.
+                2. When the file chooser appears on the screen, select any
+                 non-traversable folder from "Look-In" combo box,
+                 for example the user's folder or a folder above it.
+                 (The folder will not be opened since it's non-traversable).
+                3. Select the Documents folder again.
+                4. If NullPointerException does not occurs in the step 3,
+                 click Pass, otherwise the Test Fails.
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:77]:
```suggestion
                4. If NullPointerException does not occur in the step 3,
                 click Pass, otherwise the test fails automatically.
```
I guess this is better and has no grammar mistakes.

**Review by aivanov-jdk — APPROVED** [review]

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+                """;
+        frame = new JFrame("JFileChooser File View NPE test");
+        passFailJFrame = new PassFailJFrame("Test Instructions", INSTRUCTIONS, 5L, 13, 40);
```
Comment (kumarabhi006) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:80]: Few lines are going beyond 80 columns, can you please update them?

Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:80]: > Few lines are going beyond 80 columns, can you please update them?

I agree, line 80 could be wrapped, it's 92 columns long. At the same, I wouldn't have insisted because it contains the parameters to `PassFailJFrame` which aren't so important to understanding the logic of the test.

Line 92 with the position is better left untouched, it's not too long (five chars outside of 80-column limit) and it's cleaner unwrapped. You may statically-import `PassFailJFrame.Position` in which case the line would become shorter:
```java
        PassFailJFrame.positionTestWindow(frame, Position.HORIZONTAL);
```

I am for keeping line 61 as it is now even though it doesn't fit the limit by 5 characters. It's still readable. In fact, this `throws` clause could be avoided altogether by restructuring the code. The constructor of `PassFailJFrame` is designed to be called on the main thread, yet Tejesh prefers it this way. I don't think it's worth moving the code around at this point when it's approved already.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+import javax.swing.JFileChooser;
+import javax.swing.SwingUtilities;
+import javax.swing.WindowConstants;
+
+import javax.swing.filechooser.FileView;
+
+/*
+ * @test
+ * @bug 6616245
+ * @key headful
```
[the remark this answers is not in the document]
Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:39]: > Is it required to have a `key` tag for manual test case?

A manual test is declared by its `@run` tag:
```
 * @run main/manual FileViewNPETest
```

If `-a` (automatic tests) is passed to jtreg, this test isn't selected for running.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 17 above)
Comment (kumarabhi006) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:80]: Ok.

**Comment by kumarabhi006** [comment]
Verified the latest fix and is working fine.

**Review by kumarabhi006 — APPROVED** [review]

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 17 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Removed the tag.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 16 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Updated.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java [hunk 19]
```diff
@@ -36,7 +36,6 @@
 /*
  * @test
  * @bug 6616245
- * @key headful
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Why did you remove `headful`?

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 19 above)
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java]: Added again, though some manual tests have this tag and some don't better to keep it as an indication that represents as headful test......

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:39]: I admit I misunderstood the question. I thought @kumarabhi006 had asked for adding another keyword to `@key` to specify a _manual_ test.

I didn't think about removing `@key headful`. At the same time, a manual test implies being _headful_ to the UI with instructions.

**Review by aivanov-jdk — APPROVED** [review]

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 5 above)
Comment (mrserb) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: Probably it will be good to automate it, and cover the macOS? I do not see a reason why we cannot set Metal L&F and verify the fix.

Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: I wasn't able to reproduce the issue in macOS, so didn't include it in test...... And the fix is for metal L&F and should we explicitly set the L&F? Anyhow default L&F is metal right....?

Comment (aivanov-jdk) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: > Probably it will be good to automate it

I've submitted [JDK-8295298](https://bugs.openjdk.org/browse/JDK-8295298): Automate javax/swing/JFileChooser/FileViewNPETest.java

> … and cover the macOS? I do not see a reason why we cannot set Metal L&F and verify the fix.

Agreed.

> I wasn't able to reproduce the issue in macOS, so didn't include it in test...... And the fix is for metal L&F and should we explicitly set the L&F? Anyhow default L&F is metal right....?

The default Look and Feel is Metal except for macOS where the default one is Aqua which is the system L&F. The default L&F can be changed using a property file or setting the system property.

Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: Ok, then I will update the test and raise another PR with new bug.

Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:67]: Sorry, I had not set Metal L&F while checking in macOS, the is reproducible in macOS also, and the test case has to be updated for macOS.

File: test/jdk/javax/swing/JFileChooser/FileViewNPETest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:40]: Yeah, I checked it now and the issue is reproducible in macOS also, have to update the test to include macOS.

[the remark this answers is not in the document]
Comment (TejeshR13) [reply test/jdk/javax/swing/JFileChooser/FileViewNPETest.java:40]: Had thought that default L&F in macOS will be metal, didn't realize have to set it explicitly. My mistake, thank you @aivanov-jdk and @mrserb for pointing it out.
